misc/windows-deploy/hostsetup.ps1

```diff
@@ -11,41 +11,119 @@
 # permissions and limitations under the License.
 
+<#
+.NOTES
+This setup script is required in order to enable task IAM roles on windows instances.
+
+.NOTES
+169.254.170.2:51679 is the IP address used for task IAM roles.
+#>
+$CREDENTIAL_ADDRESS = "169.254.170.2"
+$CREDENTIAL_LISTEN_PORT = "80"
+$CREDENTIAL_CONNECT_PORT = "51679"
+
 $oldActionPref = $ErrorActionPreference
 $ErrorActionPreference = 'Continue'
 
-# This setup script is required in order to enable task IAM roles on windows instances.
-
-# 169.254.170.2:51679 is the IP address used for task IAM roles.
-$credentialAddress = "169.254.170.2"
-$credentialPort = "51679"
+$vmNetAdapter = Get-VMNetworkAdapter -SwitchName nat -ManagementOS -Name "*APIPA*" -ErrorAction:Ignore
 
-$adapter = (Get-NetAdapter -Name "*APIPA*")
-if(!($adapter)) {
-	# APIPA nat adapter controls IP range 169.254.x.x on windows.
-	Add-VMNetworkAdapter -VMNetworkAdapterName "APIPA" -SwitchName nat -ManagementOS
+if(!($vmNetAdapter)) {
+    # APIPA nat adapter controls IP range 169.254.x.x on windows.
+    Add-VMNetworkAdapter -VMNetworkAdapterName "APIPA" -SwitchName nat -ManagementOS
 }
+[int]$delay = 1
+[int]$maxTries = 10
+[bool]$available = $false
+While (-not $available -and $maxTries -ge 0) {
+    try {
+        $vmNetAdapter = Get-VMNetworkAdapter -SwitchName nat -ManagementOS -Name "*APIPA*" -ErrorAction:Ignore
+        if ($vmNetAdapter.count -ge 1) {
+            $available = $true
+        } else {
+            sleep 1
+        }
+    } catch {
+        sleep 1
+    } finally {
+        $maxTries--
+    }
+}
+[string]$vmNetAdapterMac = $vmNetAdapter.MacAddress
 
-$ifIndex = (Get-NetAdapter -Name "*APIPA*" | Sort-Object | Select ifIndex).ifIndex
+[int]$delay = 1
+[int]$maxTries = 10
+[bool]$available = $false
+While (-not $available -and $maxTries -ge 0) {
+    try {
+        $netAdapter = Get-NetAdapter | Where {$_.MacAddress.Replace('-','') -eq $vmNetAdapterMac}
+        if ($netAdapter.count -ge 1) {
+            $available = $true
+        } else {
+            sleep 1
+        }
+    } catch {
+        sleep 1
+    } finally {
+        $maxTries--
+    }
+}
 
-$dockerSubnet = (docker network inspect nat | ConvertFrom-Json).IPAM.Config.Subnet
+[int]$ifIndex = $netAdapter.ifIndex
+$dockerNatConfig = Invoke-Expression -Command "docker network inspect nat" | ConvertFrom-Json
+$dockerSubnet = $dockerNatConfig.IPAM.Config.Subnet
 
 # This address will only exist on systems that have already set up the routes.
-$ip = (Get-NetRoute -InterfaceIndex $ifIndex -DestinationPrefix $dockerSubnet)
-if(!($ip)) {
+$ip = (Get-NetRoute -InterfaceIndex $ifIndex -DestinationPrefix $dockerSubnet -ErrorAction:Ignore)
+# ip = $null when the route does not exist
 
-	# This command tells the APIPA interface that this IP exists.
-	New-NetIPAddress -InterfaceIndex $ifIndex -IPAddress $credentialAddress -PrefixLength 32
+if(!($ip)) { #TODO: Make each setup/config item idempotent rather than a simple assumption/check for the route
```

Comment:
Is this TODO in reference to the created Firewall rules and Routes added?

Reply:
Yes. How do we generally handle things like this where there is work to do... but that we are putting off until later?